### PR TITLE
example/wifi_manager: Add wifi manager sample program

### PR DESCRIPTION
**Pull request by pillip8282**
Wi-Fi control program in tash using wifi_manager APIs

### CODE REVIEW

**Comment by seinfra** [comment]
Code Rule Check OK.

File: apps/examples/wifi_manager_sample/Makefile [hunk 1]
```diff
@@ -0,0 +1,121 @@
+###########################################################################
+#
+# Copyright 2016 Samsung Electronics All Rights Reserved.
```
Comment (sunghan-chang) [inline apps/examples/wifi_manager_sample/Makefile]: 2017

File: apps/examples/wifi_manager_sample/Makefile [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+###########################################################################
+
+-include $(TOPDIR)/.config
+-include $(TOPDIR)/Make.defs
+include $(APPDIR)/Make.defs
+
+# built-in application info
+
+APPNAME = wm_test
+THREADEXEC = TASH_EXECMD_ASYNC
```
Comment (sunghan-chang) [inline apps/examples/wifi_manager_sample/Makefile]: You had better add below.
```
FUNCNAME = wm_test_main
```

File: apps/examples/wifi_manager_sample/Makefile [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+install: $(BIN_DIR)$(DELIM)$(PROGNAME)
+
+else
+install:
+
+endif
+
+ifeq ($(CONFIG_BUILTIN_APPS)$(CONFIG_EXAMPLES_WIFIMANAGER_TEST),yy)
+$(BUILTIN_REGISTRY)$(DELIM)$(APPNAME)_main.bdat: $(DEPCONFIG) Makefile
+	$(call REGISTER,$(APPNAME),$(APPNAME)_main,$(THREADEXEC))
```
Comment (sunghan-chang) [inline apps/examples/wifi_manager_sample/Makefile]:
```
$(BUILTIN_REGISTRY)$(DELIM)$(FUNCNAME).bdat: $(DEPCONFIG) Makefile
	$(call REGISTER,$(APPNAME),$(FUNCNAME),$(THREADEXEC))
```

File: apps/examples/wifi_manager_sample/README.txt [hunk 4]
```diff
@@ -0,0 +1,10 @@
+examples/ocfeasysetup
```
Comment (sunghan-chang) [inline apps/examples/wifi_manager_sample/README.txt]: Need to check

File: apps/examples/wifi_manager_sample/wm_test.c [hunk 5]
```diff
@@ -0,0 +1,625 @@
+/****************************************************************************
+ *
+ * Copyright 2016 Samsung Electronics All Rights Reserved.
```
Comment (sunghan-chang) [inline apps/examples/wifi_manager_sample/wm_test.c]: 2017

File: apps/examples/wifi_manager_sample/wm_test.c [hunk 6]
```diff
@@ -0,0 +1,625 @@
...
+#else
+	int wm_test_main(int argc, char *argv[])
+#endif
+{
+	printf("wifi manager test!!\n");
+	int res = wm_signal_init();
+	if (res < 0) {
+		return -1;
+	}
+	task_create("wifi test sample", 100, 1024 * 10, (main_t) wm_process, argv);
```
Comment (sunghan-chang) [inline apps/examples/wifi_manager_sample/wm_test.c]:
```
(main_t)wm_process
```

File: apps/examples/wifi_manager_sample/wm_test.c [hunk 7]
```diff
@@ -0,0 +1,625 @@
...
+
+	return ;
+}
+
+
+int
+wm_parse_commands(struct options *opt, int argc, char *argv[])
+{
+	if (argc < 3) {
+		return -1;
```
Comment (sunghan-chang) [inline apps/examples/wifi_manager_sample/wm_test.c]: Don't you need to support help?

File: apps/examples/wifi_manager_sample/wm_test.c [hunk 8]
```diff
@@ -0,0 +1,625 @@
...
+		printf("[pkes] Step 6 done..Disconnecting AP\n");
+		/* Disconnect AP */
+		wifi_manager_ap_disconnect();
+		WM_TEST_WAIT;
+		printf("[pkes] Step 7 done..Cycle finished [Round %d]\n", cnt);
+	}
+
+	res = wifi_manager_deinit();
+	if (res != WIFI_MANAGER_SUCCESS) {
+		printf("[pkes] WiFi Manager failed to stop\n");
```
Comment (sunghan-chang) [inline apps/examples/wifi_manager_sample/wm_test.c]: Please remove [pkes] from printf string.

File: apps/examples/wifi_manager_sample/Makefile (hunk 1 above)
Comment (pillip8282) [reply apps/examples/wifi_manager_sample/Makefile]: ok I will fix.

File: apps/examples/wifi_manager_sample/Makefile (hunk 2 above)
Comment (pillip8282) [reply apps/examples/wifi_manager_sample/Makefile]: ok I will fix.

File: apps/examples/wifi_manager_sample/Makefile (hunk 3 above)
Comment (pillip8282) [reply apps/examples/wifi_manager_sample/Makefile]: ok I will fix

File: apps/examples/wifi_manager_sample/README.txt (hunk 4 above)
Comment (pillip8282) [reply apps/examples/wifi_manager_sample/README.txt]: ok I will fix

File: apps/examples/wifi_manager_sample/wm_test.c (hunk 5 above)
Comment (pillip8282) [reply apps/examples/wifi_manager_sample/wm_test.c]: ok I will fix

File: apps/examples/wifi_manager_sample/wm_test.c (hunk 8 above)
Comment (pillip8282) [reply apps/examples/wifi_manager_sample/wm_test.c]: ok I will fix

File: apps/examples/wifi_manager_sample/wm_test.c (hunk 7 above)
Comment (pillip8282) [reply apps/examples/wifi_manager_sample/wm_test.c]: wm_process will print usage if return value is less than 0 at wm_parse_commands.

**Comment by pillip8282** [comment]
There are no wifi-control samples which call wifi-manager APIs in tizen rt. so if new wifi is ported than we need to rewrite control programs. that's the reason why it is added.

**Comment by seinfra** [comment]
Code Rule Check OK.

**Comment by seinfra** [comment]
Code Rule Check OK.

File: apps/examples/wifi_manager_sample/wm_test.c [hunk 9]
```diff
@@ -0,0 +1,661 @@
...
+	WIFI_MANAGER_AUTH_WPA_AND_WPA2_PSK,        /**<  WPA_PSK and WPA_PSK mixed mode */
+	WIFI_MANAGER_AUTH_UNKNOWN,                 /**<  unknown type                   */
+};
+
+
+static const wifi_manager_ap_crypto_type_e crypto_type_table[] = {
+	WIFI_MANAGER_CRYPTO_NONE,                  /**<  none encryption                */
+	WIFI_MANAGER_CRYPTO_UNKNOWN,               /**<  to be fixed            */
+	WIFI_MANAGER_CRYPTO_AES,                   /**<  AES encryption                 */
+	WIFI_MANAGER_CRYPTO_AES,                   /**<  AES encryption                 */
```
Comment (dr-venkman) [inline apps/examples/wifi_manager_sample/wm_test.c]: is it required to have repeated entries? are they mapped one-on-one to auth_types above?

File: apps/examples/wifi_manager_sample/wm_test.c [hunk 10]
```diff
@@ -0,0 +1,637 @@
...
+		res = wifi_manager_mode_set(STA_MODE);
+		if (res != WIFI_MANAGER_SUCCESS) {
+			printf(" Set AP mode Fail\n");
+			return ;
+		}
+		printf(" Step 4 done..Connecting to AP\n");
+		/* Connect to AP */
+		wifi_manager_ap_config_s apconfig = { "NETGEAR99", 9,
+											  "artikwifi", 9,
+											  WIFI_MANAGER_AUTH_WPA2_PSK,
```
[the remark this answers is not in the document]
Comment (pillip8282) [reply apps/examples/wifi_manager_sample/wm_test.c]: No it is not ok. I will fix.

**Comment by seinfra** [comment]
Code Rule Check OK.

File: apps/examples/wifi_manager_sample/wm_test.c [hunk 11]
```diff
@@ -0,0 +1,704 @@
...
+	wm_scan_done,
+};
+
+static pthread_mutex_t g_wm_mutex = PTHREAD_MUTEX_INITIALIZER;;
+static pthread_cond_t g_wm_cond;
+static pthread_mutex_t g_wm_func_mutex = PTHREAD_MUTEX_INITIALIZER;;
+static pthread_cond_t g_wm_func_cond;
+static int g_mode = 0; // check program is running
+
+#define SET_SOFTAP_OPTION(opt, func, ssid, pwd)	\
```
Comment (sunghan-chang) [inline apps/examples/wifi_manager_sample/wm_test.c]: This is not used.

File: apps/examples/wifi_manager_sample/wm_test.c [hunk 12]
```diff
@@ -0,0 +1,704 @@
...
+	do {										\
+		opt->func = func;						\
+		opt->ssid = ssid;						\
+		opt->password = pwd;					\
+	} while (0);
+
+#define SET_STA_OPTION(opt, func, ssid, auth, crypto, pwd)	\
```
Comment (sunghan-chang) [inline apps/examples/wifi_manager_sample/wm_test.c]: Same as previous

File: apps/examples/wifi_manager_sample/wm_test.c [hunk 13]
```diff
@@ -0,0 +1,704 @@
...
+	do {													\
+		opt->func = func;									\
+		opt->ssid = ssid;									\
+		opt->auth_type = auth;								\
+		opt->crypto_type = crypto;							\
+		opt->password = pwd;								\
+	} while (0);
+
+#define WM_TEST_TIME_DIFF												\
```
Comment (sunghan-chang) [inline apps/examples/wifi_manager_sample/wm_test.c]: Same as previous

File: apps/examples/wifi_manager_sample/wm_test.c [hunk 14]
```diff
@@ -0,0 +1,704 @@
...
+	WIFI_MANAGER_CRYPTO_AES,                   /**<  AES encryption                 */
+	WIFI_MANAGER_CRYPTO_UNKNOWN,               /**<  unknown encryption             */
+};
+
+
+wifi_manager_ap_auth_type_e
+get_auth_type(const char *method)
+{
+	int i = 0;
+	int list_size = sizeof(wifi_test_auth_method)/sizeof(wifi_test_auth_method[0]);
```
Comment (sunghan-chang) [inline apps/examples/wifi_manager_sample/wm_test.c]: You should add spaces around '/'.
```
_method) / sizeof(wifi_
```

File: apps/examples/wifi_manager_sample/wm_test.c [hunk 15]
```diff
@@ -0,0 +1,704 @@
...
+	}
+	return WIFI_MANAGER_AUTH_UNKNOWN;
+}
+
+
+wifi_manager_ap_crypto_type_e
+get_crypto_type(const char *method)
+{
+	int i = 0;
+	int list_size = sizeof(wifi_test_auth_method)/sizeof(wifi_test_auth_method[0]);
```
Comment (sunghan-chang) [inline apps/examples/wifi_manager_sample/wm_test.c]: same as previous

**Pull request merged**